misc/python/materialize/cli/ci_coverage_pr_report.py

```diff
@@ -0,0 +1,116 @@
+        ["git", "merge-base", "HEAD", base_branch], check=True, capture_output=True
+    )
+    merge_base = result.stdout.strip()
+    result = subprocess.run(
+        ["git", "diff", "-U0", merge_base], check=True, capture_output=True
+    )
+
+    coverage: Coverage = {}
+    file = None
+    for line_raw in result.stdout.splitlines():
+        line = line_raw.decode("utf-8")
+        # +++ b/src/adapter/src/coord/command_handler.rs
+        if line.startswith("+++"):
+            if not line.endswith(".rs"):
+                continue
+            file = line.removeprefix("+++ b/")
+            coverage[file] = OrderedDict()
+        # @@ -641,7 +640,6 @@ impl Coordinator {
+        elif line.startswith("@@ ") and file:
+            # We only care about the second value ("+640,6" in the example),
+            # which contains the line number and length of the modified block
+            # in new code state.
+            parts = line.split(" ")[2]
+            if "," in parts:
+                start, length = map(int, parts.split(","))
+            else:
+                start = int(parts)
+                length = 1
+            for line_nr in range(start, start + length):
+                coverage[file][line_nr] = 0
+    return coverage
+
+
+def mark_covered_lines(lcov_file: str, coverage: Coverage) -> None:
+    """
+    For a description of the lcov tracing file format, see the bottom of
+    https://linux.die.net/man/1/geninfo
+    """
+    for line in open(lcov_file):
+        line = line.strip()
+        if not line:
+            continue
+        if line == "end_of_record":
+            continue
+        method, content = tuple(line.strip().split(":", 1))
+        # SF:/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0193496e7aec9a4e3-1/materialize/coverage/src/transform/src/lib.rs
+        if method == "SF":
+            result = SOURCE_RE.search(content)
+            assert result
+            file = result.group(1)
+        # DA:111,15524
+        # DA:112,0
+        # DA:113,15901
+        elif method == "DA" and file in coverage:
+            line_str, hit_str = content.split(",", 1)
+            line_nr = int(line_str)
+            hit = int(hit_str) if hit_str.isnumeric() else int(float(hit_str))
+            if line_nr in coverage[file]:
+                coverage[file][line_nr] += hit
+
+
+def print_report(coverage: Coverage) -> None:
+    """
+    Remove uncovered lines and print a git diff
+    """
+    for file, lines in coverage.items():
+        with open(file, "r+") as f:
+            content = f.readlines()
+            f.seek(0)
+            for i, line in enumerate(content):
+                if i + 1 not in lines or lines[i + 1] > 0:
+                    f.write(line)
+            f.truncate()
+
+    subprocess.run(["git", "diff"], check=True)
+    # Restore the code into its original state
+    subprocess.run(["git", "reset", "--hard", "HEAD"], check=True, capture_output=True)
```

Comment:
Currently not necessary since nothing will be run after this. But I wanted to leave the working directory in a clean state in case other checks run after it in the future.

Reply:
Sure. But did anything get dirty at all? Or do you want to keep it in any case to be sure?

Reply:
Thanks for the explanation! I didn't spot that you were operating on the actual files for this.